### PR TITLE
Feature/controls usage

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: modules/usage/module.php [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+			return;
+		}
+
+		$document = Plugin::$instance->documents->get( $post->ID );
+
+		if ( ! $document ) {
+			return;
+		}
+
+		$is_update = 'publish' === $old_status && 'publish' === $new_status;
```
Comment (matipojo) [inline modules/usage/module.php]: Remove - It's handheld by `before_document_save`

File: modules/usage/module.php [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+	 *
+	 * @return array
+	 */
+	public function add_tracking_data( $params ) {
+		$params['usages']['elements'] = get_option( self::OPTION_NAME );
+
+		return $params;
+	}
+
+	/***
```
Comment (matipojo) [inline modules/usage/module.php]: Replace all `/***` with `/**`.

File: modules/usage/module.php [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+				$global_usage[ $doc_name ][ $element_type ] = [
+					'count' => 0,
+					'controls' => [],
+				];
+			}
+
+			$global_element_ref = &$global_usage[ $doc_name ][ $element_type ];
+			$global_element_ref['count'] += $element_data['count'];
+
+			if ( empty( $element_data['controls'] ) ) {
```
Comment (matipojo) [inline modules/usage/module.php]: Move before `$global_element_ref`.

File: modules/usage/module.php [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+	/***
+	 * Remove from global
+	 *
+	 * Remove's usage from global (update database).
+	 *
+	 * @param Document $document
+	 */
+	private function remove_from_global( $document ) {
+		$prev_usage = $document->get_meta( self::META_KEY );
```
Comment (matipojo) [inline modules/usage/module.php]: `if ( empty( $prev_usage ) ) { return; }`

File: modules/usage/module.php [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+
+				$tab = $control_config['tab'];
+
+				if ( ! isset( $control_config['section'], $control_config['default'] ) ) {
+					continue;
+				}
+
+				$section = $control_config['section'];
+
+				// If setting value is not the control default
```
Comment (matipojo) [inline modules/usage/module.php]: Add a `.` in the end of each comment.

File: modules/usage/module.php [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+	 * Save requested document usage, and update global.
+	 *
+	 * @param Document $document
+	 */
+	private function save_document_usage( Document $document ) {
+		if ( ! $document::get_property( 'is_editable' ) ) {
+			return;
+		}
+
+		// current
```
Comment (matipojo) [inline modules/usage/module.php]: Capital

File: modules/usage/module.php [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+
+	/**
+	 * Usage module constructor.
+	 *
+	 * Initializing Elementor usage module.
+	 *
+	 * @access public
+	 */
+	public function __construct() {
+		add_action( 'transition_post_status', [ $this, 'on_status_change' ], 10, 3 );
```
Comment (matipojo) [inline modules/usage/module.php]: On upgrade: Needs to remove the old usage format from post meta.

File: modules/usage/usage-reporter.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+ * the user.
+ *
+ * @since 1.0.0
+ */
+class Usage_Reporter extends Base_Reporter {
+
+	public function get_title() {
+		$title = 'Elements Usage';
+
+		if ( empty( $_GET['elementor_usage_recalc'] ) ) { // phpcs:ignore -- nonce validation is not require here.
```
Comment (matipojo) [inline modules/usage/usage-reporter.php]: Extract `elementor_usage_recalc` to a const.

File: package.json [hunk 9]
```diff
@@ -59,4 +59,4 @@
 		"react": "^16.8.6",
 		"react-dom": "^16.8.6"
 	}
-}
+}
```
Comment (matipojo) [inline package.json]: Add an empty line.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/usage/module.php [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+
+		add_action( 'elementor/document/before_save', [ $this, 'before_document_save' ] );
```
Comment (matipojo) [inline modules/usage/module.php]: Check id there are errors on save a local template

File: modules/usage/module.php (hunk 1 above)
Comment (iNewLegend) [reply modules/usage/module.php]:
```suggestion
		$is_update = 'publish' === $old_status && 'publish' === $new_status;
```

File: modules/usage/module.php [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+	/**
+	 * Recalculate usage.
+	 *
+	 * Recalculate usage for all elementor posts.
+	 *
+	 * @param int $limit
+	 * @param int $offset
+	 */
+	public function recalc_usage( $limit = -1, $offset = 0 ) {
+		if ( 0 === $offset ) {
```
Comment (matipojo) [inline modules/usage/module.php]: Explain

File: modules/usage/module.php [hunk 12]
```diff
@@ -0,0 +1,517 @@
...
+			delete_option( self::OPTION_NAME );
+		}
+
+		$post_types = get_post_types( array( 'public' => true ) );
+
+		$query = new WP_Query( [
+			'meta_key' => '_elementor_data',
+			'post_type' => $post_types,
+			'post_status' => 'publish',
```
Comment (matipojo) [inline modules/usage/module.php]:
```suggestion
			'post_status' => [ 'publish', 'private' ],
```

File: modules/usage/module.php [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+			$control_config = $element_controls[ $control ];
+
+			if ( ! isset( $control_config['section'], $control_config['default'] ) ) {
+				continue;
+			}
+
+			$tab = $control_config['tab'];
+			$section = $control_config['section'];
+
+			// If setting value is not the control default.
```
Comment (matipojo) [inline modules/usage/module.php]:
```suggestion
			// If the setting value is equal to control default.
```

File: modules/usage/usage-reporter.php [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Elementor usage report.
+ *
+ * Elementor system report handler class responsible for generating a report for
+ * the user.
+ *
+ * @since 1.0.0
```
Comment (matipojo) [inline modules/usage/usage-reporter.php]: Remove

File: modules/usage/usage-reporter.php [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+		$usage = '<tr>';
+
+		foreach ( $module->get_formatted_usage() as $doc_type => $data ) {
+			$usage .= '<td>' . $data['title'] . '</td><td>';
+
+			foreach ( $data['elements'] as $element => $count ) {
+				$usage .= $element . ': ' . $count . PHP_EOL;
+			}
+
+			$usage .= '</td></tr>';
```
Comment (matipojo) [inline modules/usage/usage-reporter.php]: The `tr` is closed twice

File: tests/phpunit/elementor/modules/usage/dynamic-tags/link.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	public function get_content_type() {
+		return '';
+	}
+
+	public function get_name() {
+		return 'post-url';
+	}
+}
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/usage/dynamic-tags/link.php]: Add empty line

File: tests/phpunit/elementor/modules/usage/dynamic-tags/title.php [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	public function get_content_type() {
+		return '';
+	}
+
+	public function get_name() {
+		return 'post-title';
+	}
+}
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/usage/dynamic-tags/title.php]: Add empty line

File: modules/usage/module.php [hunk 18]
```diff
@@ -0,0 +1,517 @@
...
+
+			if ( ! $element_instance ) {
+				return $element;
+			}
+
+			$element_controls = $element_instance->get_controls();
+			$settings_controls = $element['settings'];
+			$element_ref = &$usage[ $type ];
+
+			$settings_controls = $this->add_general_controls( $settings_controls, $element_ref );
```
Comment (matipojo) [inline modules/usage/module.php]:
```suggestion
			// Add dynamic values.
			$settings_controls = $this->add_general_controls( $settings_controls, $element_ref );
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/upgrade/upgrades.php [hunk 19]
```diff
@@ -555,4 +555,12 @@ public static function _v_2_6_6_fa4_migration_button( $updater ) {
 		Upgrade_Utils::_update_widget_settings( 'button', $updater, $changes );
 		Upgrade_Utils::_update_widget_settings( 'icon-box', $updater, $changes );
 	}
+
+	/**
+	 * Format was changed.
+	 */
+	public static function _v_2_6_6_remove_old_usage_data() {
+		delete_option( \Elementor\Modules\Usage\Module::OPTION_NAME );
+		delete_post_meta_by_key( \Elementor\Modules\Usage\Module::META_KEY );
```
Comment (matipojo) [inline core/upgrade/upgrades.php]: Add recalc()

**Review by matipojo — APPROVED** [review]

File: core/upgrade/upgrades.php [hunk 20]
```diff
@@ -574,4 +575,24 @@ public static function _v_2_6_7_rename_document_types_to_wp() {
...
+	/**
+	 * Recalc usage.
+	 *
+	 * @param Updater $updater
+	 */
+	public static function _v_2_7_0_recalc_usage_data( $updater ) {
+		/** @var Module $module */
+		$module = Plugin::$instance->modules_manager->get_modules( 'usage' );
+
+		$module->recalc_usage( $updater->get_limit(), $updater->get_current_offset() );
```
Comment (KingYes) [inline core/upgrade/upgrades.php]: check success?

File: modules/usage/module.php [hunk 21]
```diff
@@ -0,0 +1,520 @@
+<?php
+
```
Comment (KingYes) [inline modules/usage/module.php]: remove line

File: modules/usage/module.php [hunk 22]
```diff
@@ -0,0 +1,520 @@
...
+namespace Elementor\Modules\Usage;
+
+use Elementor\Core\Base\Document;
+use Elementor\Core\Base\Module as BaseModule;
+use Elementor\Core\DynamicTags\Manager;
+use Elementor\System_Info\Main as System_Info;
+use Elementor\DB;
+use Elementor\Plugin;
+use WP_Post;
```
Comment (KingYes) [inline modules/usage/module.php]: Use backslash instead

File: modules/usage/module.php [hunk 23]
```diff
@@ -0,0 +1,520 @@
...
+			}
+
+			$tab = $control_config['tab'];
+			$section = $control_config['section'];
+
+			// If setting value is not the control default.
+			if ( $value !== $control_config['default'] ) {
+				$this->increase_controls_count( $element_ref, $tab, $section, $control, 1 );
+
+				$changed_controls_count ++;
```
Comment (KingYes) [inline modules/usage/module.php]: remove space

File: modules/usage/module.php [hunk 24]
```diff
@@ -0,0 +1,520 @@
...
+	 *
+	 * Increase controls count, for each element.
+	 *
+	 * @param array & $element_ref
+	 * @param string $tab
+	 * @param string $section
+	 * @param string $control
+	 * @param int $count
+	 */
+	private function increase_controls_count( & $element_ref, $tab, $section, $control, $count ) {
```
Comment (KingYes) [inline modules/usage/module.php]: remove space

File: modules/usage/module.php [hunk 25]
```diff
@@ -0,0 +1,520 @@
...
+	 *
+	 * Add's controls to this element_ref, returns changed controls count.
+	 *
+	 * @param array $settings_controls
+	 * @param array $element_controls
+	 * @param array & $element_ref
+	 *
+	 * @return int ($changed_controls_count).
+	 */
+	private function add_controls( $settings_controls, $element_controls, & $element_ref ) {
```
Comment (KingYes) [inline modules/usage/module.php]: remove space

File: modules/usage/module.php [hunk 26]
```diff
@@ -0,0 +1,520 @@
...
+	 * Add general controls.
+	 *
+	 * Extract general controls to element ref, return clean `$settings_control`.
+	 *
+	 * @param array $settings_controls
+	 * @param array & $element_ref
+	 *
+	 * @return array ($settings_controls).
+	 */
+	private function add_general_controls( $settings_controls, & $element_ref ) {
```
Comment (KingYes) [inline modules/usage/module.php]: remove space

File: modules/usage/module.php [hunk 27]
```diff
@@ -0,0 +1,520 @@
...
+		if ( ! empty( $settings_controls[ Manager::DYNAMIC_SETTING_KEY ] ) ) {
+			$settings_controls = array_merge( $settings_controls, $settings_controls[ Manager::DYNAMIC_SETTING_KEY ] );
+
+			// Add dynamic count to controls under `global` tab.
```
Comment (KingYes) [inline modules/usage/module.php]:
```suggestion
			// Add dynamic count to controls under `general` tab.
```

File: modules/usage/module.php [hunk 28]
```diff
@@ -0,0 +1,520 @@
...
+
+			if ( ! isset( $usage[ $type ] ) ) {
+				$usage[ $type ] = [
+					'count' => 0,
+					'control_percent' => 0,
+					'controls' => [],
+				];
+			}
+
+			$usage[ $type ]['count'] ++;
```
Comment (KingYes) [inline modules/usage/module.php]:
```suggestion
			$usage[ $type ]['count']++;
```

File: modules/usage/module.php [hunk 29]
```diff
@@ -0,0 +1,520 @@
...
+	public function __construct() {
+		add_action( 'transition_post_status', [ $this, 'on_status_change' ], 10, 3 );
+		add_action( 'before_delete_post', [ $this, 'on_before_delete_post' ] );
+
+		add_action( 'elementor/document/before_save', [ $this, 'before_document_save' ] );
+		add_action( 'elementor/document/after_save', [ $this, 'after_document_save' ] );
+
+		add_filter( 'elementor/tracker/send_tracking_data_params', [ $this, 'add_tracking_data' ] );
+
+		System_Info::add_report( 'usage', [
```
Comment (KingYes) [inline modules/usage/module.php]: on admin_init

File: modules/usage/usage-reporter.php [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+ * the user.
+ */
+class Usage_Reporter extends Base_Reporter {
+
+	const RECALC_ACTION = 'elementor_usage_recalc';
+
+	public function get_title() {
+		$title = 'Elements Usage';
+
+		if ( empty( $_GET[ self::RECALC_ACTION ] ) ) { // phpcs:ignore -- nonce validation is not require here.
```
Comment (KingYes) [inline modules/usage/usage-reporter.php]: check raw

**Review by KingYes — CHANGES_REQUESTED** [review]

File: modules/usage/module.php (hunk 25 above)
Comment (matipojo) [reply modules/usage/module.php]: @iNewLegend

File: modules/usage/module.php [hunk 31]
```diff
@@ -8,7 +7,6 @@
 use Elementor\System_Info\Main as System_Info;
 use Elementor\DB;
 use Elementor\Plugin;
-use WP_Post;
 use WP_Query;
```
Comment (matipojo) [inline modules/usage/module.php]: Remove (Use backslash)

File: modules/usage/usage-reporter.php [hunk 32]
```diff
@@ -20,7 +20,7 @@ class Usage_Reporter extends Base_Reporter {
 	public function get_title() {
 		$title = 'Elements Usage';
 
-		if ( empty( $_GET[ self::RECALC_ACTION ] ) ) { // phpcs:ignore -- nonce validation is not require here.
+		if ( 'raw' !== $this->_properties['format'] && empty( $_GET[ self::RECALC_ACTION ] ) ) { // phpcs:ignore -- nonce validation is not require here.
```
Comment (matipojo) [inline modules/usage/usage-reporter.php]: if html == ...

Comment (iNewLegend) [reply modules/usage/usage-reporter.php]: for html the format is empty.

**Review by KingYes — APPROVED** [review]

**Pull request merged**